src/Illuminate/Validation/Rules/File.php

```diff
@@ -208,26 +230,131 @@ public function max($size)
     /**
      * Convert a potentially human-friendly file size to kilobytes.
      *
-     * @param  string|int  $size
-     * @return mixed
+     * Supports suffix detection with precedence over instance settings.
      */
-    protected function toKilobytes($size)
+    protected function toKilobytes(string|int $size): float|int
     {
         if (! is_string($size)) {
             return $size;
         }
 
-        $size = strtolower(trim($size));
+        if (($value = $this->parseSize($size)) === false || $value < 0) {
+            throw new InvalidArgumentException('Invalid numeric value in file size.');
+        }
+
+        return $this->detectUnits($size) === self::BINARY
+            ? $this->toBinaryKilobytes($size, $value)
+            : $this->toInternationalKilobytes($size, $value);
+    }
+
+    /**
+     * Parse the numeric portion from a file size string.
+     */
+    protected function parseSize($size): false|float
+    {
+        return filter_var(
+            is_numeric($size)
+                ? $size
+                : Str::before(trim($size), Str::match('/[a-zA-Z]/', trim($size))),
+            FILTER_VALIDATE_FLOAT, FILTER_FLAG_ALLOW_THOUSAND
+        );
+    }
+
+    /**
+     * Detect the suffix and determine appropriate units from a file size string.
+     */
+    protected function detectUnits(string $size): string
+    {
+        return match (true) {
+            is_numeric($size) => $this->units,
+            in_array(strtolower(substr(trim($size), -3)), ['kib', 'mib', 'gib', 'tib']) => self::BINARY,
+            in_array(strtolower(substr(trim($size), -2)), ['kb', 'mb', 'gb', 'tb']) => self::INTERNATIONAL,
```

Comment:
Wouldn't this be "cleaner"?
```suggestion
        $size = strtolowercase(trim($size));

        return match (true) {
            is_numeric($size) => $this->units,
            str_ends_with(), 'kib'),
            str_ends_with($size), 'mib'),
            str_ends_with($size), 'gib'),
            str_ends_with($size), 'tib') => self::BINARY,
            str_ends_with($size), 'kb'),
            str_ends_with($size), 'mb'),
            str_ends_with($size), 'gb'),
            str_ends_with($size), 'tb') => self::INTERNATIONAL,
```
```